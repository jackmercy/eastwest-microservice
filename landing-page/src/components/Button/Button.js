import React, { memo, forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import useStyles from './button.styles';

const Button = forwardRef(
  (
    {
      tag,
      size,
      color,
      disabled,
      children,
      dataTest,
      className,
      keepSizeOnMobile,
      ...restProps
    },
    ref,
  ) => {
    const classes = useStyles();
    // Have to to this to fix error
    // React.createElement: type is invalid -- expected a string or a class/function but got: undefined
    const Component = `${tag}`;

    return (
      <Component
        ref={ref}
        className={classNames(
          classes.root,
          size,
          color,
          disabled && 'disabled',
          keepSizeOnMobile && 'keepSize',
          className,
        )}
        disabled={disabled}
        data-test={dataTest || 'component-button-root'}
        {...restProps}
      >
        {children}
      </Component>
    );
  },
);
Button.propTypes = {
  children: PropTypes.node.isRequired,
  tag: PropTypes.string,
  disabled: PropTypes.bool,
  dataTest: PropTypes.string,
  className: PropTypes.string,
  keepSizeOnMobile: PropTypes.bool,
  size: PropTypes.oneOf(['large', 'medium', 'small']),
  color: PropTypes.oneOf(['primary', 'secondary', 'tertiary']),
};
Button.defaultProps = {
  dataTest: '',
  className: '',
  tag: 'button',
  size: 'medium',
  disabled: false,
  color: 'primary',
  keepSizeOnMobile: false,
};

export default memo(Button);
