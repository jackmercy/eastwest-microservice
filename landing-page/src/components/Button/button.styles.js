import { createUseStyles } from 'react-jss';

import { rem } from '../../utils/jss';
import COLORS from '../../constants/COLORS';
import BREAKPOINTS from '../../constants/BREAKPOINTS';

export default createUseStyles({
  root: {
    fontFamily: 'inherit',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    fontWeight: 600,
    lineHeight: rem(24),
    padding: [0, rem(16)],
    textAlign: 'center',
    borderRadius: rem(5),
    position: 'relative',
    transition: 'background-color .1s linear',
    whiteSpace: 'nowrap',
    userSelect: 'none',
    textDecoration: 'none',
    boxSizing: 'border-box',
    overflow: 'hidden',
    '&:focus': { outline: 'none' },
    '&:hover': { cursor: 'pointer' },
    '&.large': { height: rem(48), fontSize: rem(18) },
    '&.medium': { height: rem(40), fontSize: rem(16) },
    '&.small': { height: rem(32), fontSize: rem(14) },
    '&.primary': {
      border: 'none',
      color: COLORS.textWhite,
      backgroundColor: COLORS.primary,
      boxShadow: '2px 4px 12px rgba(0, 0, 0, 0.15)',
      '&.disabled': {
        boxShadow: 'none',
        cursor: 'not-allowed',
        color: COLORS.textGray,
        backgroundColor: COLORS.bgGray02,
      },
    },
    '&.secondary': {
      color: COLORS.primary,
      border: `1px solid ${COLORS.primary}`,
      backgroundColor: COLORS.bgWhite,
      '&.disabled': {
        cursor: 'not-allowed',
        borderColor: COLORS.borderGray01,
        color: COLORS.textGray,
      },
    },
    '&.tertiary': {
      color: COLORS.textBlack01,
      border: 'none',
      backgroundColor: 'transparent',
      '&.disabled': { cursor: 'not-allowed', color: COLORS.textGray },
    },
  },

  '@media (hover: hover) and (pointer: fine)': {
    root: {
      '&.primary:not(.disabled):hover': {
        opacity: 0.75,
      },
      '&.secondary:not(.disabled):hover': {
        opacity: 0.75,
      },
      '&.tertiary:not(.disabled):hover': {
        opacity: 0.75,
      },
    },
  },
  [`@media only screen and (max-width: ${BREAKPOINTS.maxSm}px)`]: {
    root: {
      '&:not(.keepSize)': {
        '&.large, &.medium, &.small': { fontSize: rem(14), height: rem(32) },
      },
    },
  },
});
