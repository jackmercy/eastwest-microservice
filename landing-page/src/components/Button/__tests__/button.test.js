import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Button from '../Button';

describe('Testing Button component', () => {
  it('should render match snapshot', () => {
    const wrapper = shallow(<Button tag="button">Button</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
