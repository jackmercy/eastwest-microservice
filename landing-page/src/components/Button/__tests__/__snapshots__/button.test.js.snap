// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Button component should render match snapshot 1`] = `
<button
  className="root-0-2-1"
  data-test="component-button-root"
>
  Button
</button>
`;
